File: bifrost-py/src/bifrost/ejson/base64.py
"""Base64 encoder/decoder — port of ``bifrost-ts/src/ejson/base64.ts``.

The encoding alphabet MUST be the standard MIME alphabet
(``A-Za-z0-9+/``). The decoder additionally accepts URL-safe input
(``-`` and ``_``) as described in PROTOCOL.md §3.2.

We intentionally avoid the stdlib ``base64`` module so that behavior
stays byte-identical to the TS port — stdlib is stricter about
padding, URL-safe handling, and whitespace than the TS implementation.
"""

from __future__ import annotations

BASE64_ALPHABET = (
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
)

_INVALID_CHAR = 0xFF

_LOOKUP = bytearray([_INVALID_CHAR] * 256)
for _i, _c in enumerate(BASE64_ALPHABET):
    _LOOKUP[ord(_c)] = _i
# Accept URL-safe input as well.
_LOOKUP[ord("-")] = 62
_LOOKUP[ord("_")] = 63


def _read_char(value: str, index: int) -> int:
    decoded = _LOOKUP[ord(value[index])]
    if decoded == _INVALID_CHAR:
        raise ValueError("Invalid base64 character")
    return decoded


def encode_base64(data: bytes) -> str:
    """Encode ``bytes`` as base64 (standard alphabet, ``=`` padded)."""

    out: list[str] = []
    length = len(data)

    for index in range(0, length, 3):
        byte0 = data[index] if index < length else 0
        byte1 = data[index + 1] if index + 1 < length else 0
        byte2 = data[index + 2] if index + 2 < length else 0
        chunk = (byte0 << 16) | (byte1 << 8) | byte2
        remaining = length - index

        out.append(BASE64_ALPHABET[(chunk >> 18) & 63])
        out.append(BASE64_ALPHABET[(chunk >> 12) & 63])
        out.append(BASE64_ALPHABET[(chunk >> 6) & 63] if remaining > 1 else "=")
        out.append(BASE64_ALPHABET[chunk & 63] if remaining > 2 else "=")

    return "".join(out)


def decode_base64(value: str) -> bytes:
    """Decode a base64 string. Accepts both standard and URL-safe alphabets."""

    if len(value) % 4 != 0:
        raise ValueError("Invalid base64 string length")

    pad_index = value.find("=")
    valid_length = len(value) if pad_index == -1 else pad_index
    placeholder_length = 0 if valid_length == len(value) else 4 - (valid_length % 4)

    output_length = (valid_length + placeholder_length) * 3 // 4 - placeholder_length
    out = bytearray(output_length)

    byte_index = 0
    end = valid_length - 4 if placeholder_length > 0 else valid_length
    index = 0
    while index < end:
        chunk = (
            (_read_char(value, index) << 18)
            | (_read_char(value, index + 1) << 12)
            | (_read_char(value, index + 2) << 6)
            | _read_char(value, index + 3)
        )
        out[byte_index] = (chunk >> 16) & 0xFF
        out[byte_index + 1] = (chunk >> 8) & 0xFF
        out[byte_index + 2] = chunk & 0xFF
        byte_index += 3
        index += 4

    if placeholder_length == 2:
        chunk = (_read_char(value, index) << 2) | (
            _read_char(value, index + 1) >> 4
        )
        out[byte_index] = chunk & 0xFF
    elif placeholder_length == 1:
        chunk = (
            (_read_char(value, index) << 10)
            | (_read_char(value, index + 1) << 4)
            | (_read_char(value, index + 2) >> 2)
        )
        out[byte_index] = (chunk >> 8) & 0xFF
        out[byte_index + 1] = chunk & 0xFF

    return bytes(out)

File: bifrost-ts/src/ejson/base64.ts
import { newBinary } from './utils'

const BASE64_ALPHABET =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'

const INVALID_CHAR = 255

const BASE64_LOOKUP = (() => {
  const lookup = new Uint8Array(256)
  lookup.fill(INVALID_CHAR)

  for (const [index, char] of [...BASE64_ALPHABET].entries()) {
    lookup[char.charCodeAt(0)] = index
  }

  // Accept URL-safe input as well.
  lookup['-'.charCodeAt(0)] = 62
  lookup['_'.charCodeAt(0)] = 63

  return lookup
})()

function getBase64Lengths(value: string): {
  validLength: number
  placeholderLength: number
} {
  if (value.length % 4 !== 0) {
    throw new Error('Invalid base64 string length')
  }

  const validLength = value.indexOf('=')
  const normalizedValidLength = validLength === -1 ? value.length : validLength
  const placeholderLength =
    normalizedValidLength === value.length ? 0 : 4 - (normalizedValidLength % 4)

  return {
    validLength: normalizedValidLength,
    placeholderLength,
  }
}

function readBase64Char(value: string, index: number): number {
  const decoded = BASE64_LOOKUP[value.charCodeAt(index)]

  if (decoded === INVALID_CHAR) {
    throw new Error('Invalid base64 character')
  }

  return decoded
}

/**
 * Encodes binary payloads for EJSON without relying on a CJS-only dependency
 * that breaks browser ESM loading in source-first consumers.
 */
export function encodeBase64(bytes: Uint8Array): string {
  let output = ''

  for (let index = 0; index < bytes.length; index += 3) {
    const byte0 = bytes[index] ?? 0
    const byte1 = bytes[index + 1] ?? 0
    const byte2 = bytes[index + 2] ?? 0
    const chunk = (byte0 << 16) | (byte1 << 8) | byte2
    const remaining = bytes.length - index

    output += BASE64_ALPHABET[(chunk >> 18) & 63]
    output += BASE64_ALPHABET[(chunk >> 12) & 63]
    output += remaining > 1 ? BASE64_ALPHABET[(chunk >> 6) & 63] : '='
    output += remaining > 2 ? BASE64_ALPHABET[chunk & 63] : '='
  }

  return output
}

/**
 * Decodes EJSON base64 payloads while preserving the legacy polyfill
 * behavior for runtimes without `Uint8Array`.
 */
export function decodeBase64(value: string): Uint8Array | ReturnType<typeof newBinary> {
  const { validLength, placeholderLength } = getBase64Lengths(value)
  const outputLength =
    ((validLength + placeholderLength) * 3) / 4 - placeholderLength
  const bytes = newBinary(outputLength)

  let byteIndex = 0
  const end = placeholderLength > 0 ? validLength - 4 : validLength
  let index = 0

  for (; index < end; index += 4) {
    const chunk =
      (readBase64Char(value, index) << 18) |
      (readBase64Char(value, index + 1) << 12) |
      (readBase64Char(value, index + 2) << 6) |
      readBase64Char(value, index + 3)

    bytes[byteIndex++] = (chunk >> 16) & 255
    bytes[byteIndex++] = (chunk >> 8) & 255
    bytes[byteIndex++] = chunk & 255
  }

  if (placeholderLength === 2) {
    const chunk =
      (readBase64Char(value, index) << 2) |
      (readBase64Char(value, index + 1) >> 4)

    bytes[byteIndex++] = chunk & 255
  }

  if (placeholderLength === 1) {
    const chunk =
      (readBase64Char(value, index) << 10) |
      (readBase64Char(value, index + 1) << 4) |
      (readBase64Char(value, index + 2) >> 2)

    bytes[byteIndex++] = (chunk >> 8) & 255
    bytes[byteIndex] = chunk & 255
  }

  return bytes
}
